### PR TITLE
feat: merge cosmos-sdk v0.45.16

### PR DESCRIPTION
**Pull request by JimLarson**
## Description

Refs: Agoric/agoric-sdk#8223

Merges [cosmos/cosmos-sdk v0.45.16](https://github.com/cosmos/cosmos-sdk/releases/tag/v0.45.16).

See commit for description of merge conflict resolutions. Pay particular attention to `baseapp/abci.go`.

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] added `!` to the type prefix if API or client breaking change
* [x] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [x] provided a link to the relevant issue or specification
* [x] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/build/building-modules/01-intro.md)
* [x] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [x] added a changelog entry to `CHANGELOG.md`
* [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
* [x] updated the relevant documentation or specification
* [x] reviewed "Files changed" and left comments if necessary
* [x] run `make lint` and `make test`
* [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed all author checklist items have been addressed 
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage
* [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by JimLarson** [comment]
@mhofman please take a look at the merge of `snapshots/manager_test.go` and`store/rootmulti/snapshot_test.go` which have some opaque binary data for tests. Our changes confict with upstream changes, and the tests produce our expected values, not theirs. Please confirm that this is the appropriate merge.

**Comment by mhofman** [comment]
Reviewing nit: The switch to `agoric-labs/cometbft` being done as part of the conflict resolution of the tag merge makes it harder to follow what happened during that merge. Manually updating the version first then merging may make this review easier. I do see that `tendermint` was updated a couple times first upstream before the switch, so in addition or alternatively staggering multiple merges might also make it easier to review.

**Review by michaelfig — APPROVED** [review]
I was able to find my way with `git log -p --cc 7ec432f1e`.  LGTM!

**Review by mhofman — COMMENTED** [review]
@michaelfig's command indeed showed only the changes from conflict resolution, not sure why GH didn't manage to do the same (my experience is that it can normally show a condensed diff of such merges)

Anyway, I have replicated this merge step by step and arrived at a similar outcome. (I tried to similarly replicate the cometbft 0.34.27 merge but that caused a flurry of merge conflicts I didn't understand and didn't feel like tracking down)

I believe that unlike upstream we're not impacted by the hash changes in snapshots when switching to cometbft because we backported a fix from 0.47 that insulated us from it. See https://github.com/agoric-labs/cosmos-sdk/pull/297/commits/17f110354cf9a3c99ec60975c28af7b3e0b720dc

**Review by mhofman — APPROVED** [review]

File: CHANGELOG-Agoric.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+"Client Breaking" for breaking Protobuf, gRPC and REST routes used by end-users.
+"CLI Breaking" for breaking CLI commands.
+"API Breaking" for breaking exported APIs used by developers building on SDK.
+"State Machine Breaking" for any changes that result in a different AppState given same genesisState and txList.
+Ref: https://keepachangelog.com/en/1.0.0/
+-->
+
+# Changelog
+
+## [v0.46.16-alpha.agoric.1] - 2023-09-15
```
Comment (mhofman) [inline CHANGELOG-Agoric.md]: Nit: this is assuming we're tagging this branch or the resulting merge commit of this PR. Why not use the `## [Unreleased]` convention?

Comment (JimLarson) [reply CHANGELOG-Agoric.md]: Quite right - done.

**Pull request merged**